internal/core/block/block.go

```diff
@@ -0,0 +1,342 @@
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package coreblock
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/codec/dagcbor"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/node/bindnode"
+	"github.com/ipld/go-ipld-prime/schema"
+	"github.com/multiformats/go-multicodec"
+
+	"github.com/sourcenetwork/defradb/internal/core"
+	"github.com/sourcenetwork/defradb/internal/core/crdt"
+)
+
+// Schema is the IPLD schema type that represents a `Block`.
+var Schema schema.Type
+
+func init() {
+	Schema = mustSetSchema()
+}
+
+func mustSetSchema() schema.Type {
+	ts, err := ipld.LoadSchemaBytes([]byte(`
+		type Block struct {
+			delta	CRDT
+			links	[ DAGLink ]
+		}
+		
+		type DAGLink struct { 
+			name	String
+			link 	Link
+		}
+
+		type CRDT union {
+			| LWWRegDelta "lww"
+			| CompositeDAGDelta "composite"
+			| CounterDeltaInt "counterInt"
+			| CounterDeltaFloat "counterFloat"
+		} representation keyed
+
+		type LWWRegDelta struct {
+			docID     		Bytes
+			fieldName 		String
+			priority  		Int
+			schemaVersionID String
+			data            Bytes
+		}
+
+		type CompositeDAGDelta struct {
+			docID     		Bytes
+			fieldName 		String
+			priority  		Int
+			schemaVersionID String
+			status          Int
+		}
+
+		type CounterDeltaFloat struct {
+			docID     		Bytes
+			fieldName 		String
+			priority  		Int
+			nonce 			Int
+			schemaVersionID String
+			data            Float
+		}
+
+		type CounterDeltaInt struct {
+			docID     		Bytes
+			fieldName 		String
+			priority  		Int
+			nonce 			Int
+			schemaVersionID String
+			data            Int
+		}
+
+	`))
+	if err != nil {
+		panic(err)
+	}
+	return ts.TypeByName("Block")
+}
+
+// DAGLink represents a link to another object in a DAG.
+type DAGLink struct {
+	// Name is the name of the link.
+	//
+	// This will be either the field name of the CRDT delta or "_head" for the head link.
+	Name string
+	// Link is the CID link to the object.
+	cidlink.Link
+}
+
+func NewDAGLink(name string, link cidlink.Link) DAGLink {
+	return DAGLink{
+		Name: name,
+		Link: link,
+	}
+}
+
+// Block is a block that contains a CRDT delta and links to other blocks.
+type Block struct {
+	// Delta is the CRDT delta that is stored in the block.
+	Delta CRDT
+	// Links are the links to other blocks in the DAG.
+	Links []DAGLink
+}
+
+// New creates a new block with the given delta and links.
+func New(delta core.Delta, links []DAGLink, heads ...cid.Cid) *Block {
+	blockLinks := make([]DAGLink, 0, len(links)+len(heads))
+	for _, head := range heads {
+		blockLinks = append(
+			blockLinks,
+			DAGLink{
+				Name: "_head",
```

Comment:
question: Was this renamed from `head` in this PR, or am I misreading the deleted code?

Reply:
It was not renamed. I'm curious what makes you think that.

Reply:
ln 219 of the old internal/core/crdt/composite.go had a `"head"` string, that might have been a bug or dead code though, as everywhere else seems to say `"_head"`.  If you've copied that line somewhere (maybe to here, I have forgotten) - can you try and see if we can remove it?

Reply:
The `DeltaDecode` method that you are referring to was only used in the versioned fetcher and if I think that if that line actually worked, it would have been a bug because the next function that was called was `ProcessNode` and it needed the `_head` links to work properly. Adding the `_head` links here cannot be removed.

Reply:
Nice, thanks for the explanation :)